rust/main/agents/relayer/src/msg/op_submitter.rs

```diff
@@ -342,86 +371,247 @@ async fn prepare_task(
-                );
-                // The submit queue is too long, so give some time before checking again
-                sleep(Duration::from_millis(150)).await;
-                continue;
-            }
+        if apply_backpressure(&submit_queue, &max_submit_queue_len).await {
+            continue;
         }
-        // Pop messages here according to the configured batch.
-        let mut batch = prepare_queue.pop_many(ops_to_prepare).await;
-        if batch.is_empty() {
-            // queue is empty so give some time before checking again to prevent burning CPU
-            sleep(Duration::from_millis(100)).await;
+
+        let batch = if let Some(batch) = get_batch_or_wait(&mut prepare_queue, max_batch_size).await
+        {
+            batch
+        } else {
+            continue;
+        };
+
+        process_batch(
+            domain.clone(),
+            batch,
+            &mut prepare_queue,
+            &submit_queue,
+            &confirm_queue,
+            &metrics,
+        )
+        .await;
+    }
+}
+
+#[instrument(skip_all, fields(%domain))]
+#[allow(clippy::too_many_arguments)]
+async fn prepare_lander_task(
+    entrypoint: Arc<PayloadDispatcherEntrypoint>,
+    domain: HyperlaneDomain,
+    mut prepare_queue: OpQueue,
+    submit_queue: OpQueue,
+    confirm_queue: OpQueue,
+    max_batch_size: u32,
+    max_submit_queue_len: Option<u32>,
+    metrics: SerialSubmitterMetrics,
+    db: Arc<dyn HyperlaneDb>,
+) {
+    loop {
+        if apply_backpressure(&submit_queue, &max_submit_queue_len).await {
             continue;
         }
-        let mut task_prep_futures = vec![];
-        let op_refs = batch.iter_mut().map(|op| op.as_mut()).collect::<Vec<_>>();
-        for op in op_refs {
-            trace!(?op, "Preparing operation");
-            debug_assert_eq!(*op.destination_domain(), domain);
-            task_prep_futures.push(op.prepare());
+
+        let batch = if let Some(batch) = get_batch_or_wait(&mut prepare_queue, max_batch_size).await
+        {
+            batch
+        } else {
+            continue;
+        };
+
+        let batch_to_process = filter_ops_to_prepare(
+            entrypoint.clone(),
+            &prepare_queue,
+            &confirm_queue,
+            db.clone(),
+            batch,
+        )
+        .await;
+
+        process_batch(
+            domain.clone(),
+            batch_to_process,
+            &mut prepare_queue,
+            &submit_queue,
+            &confirm_queue,
+            &metrics,
+        )
+        .await;
+    }
+}
+
+/// This function checks the status of the payloads associated with the operations in the batch.
+/// If the payload is dropped or does not exist, the operation will go through prepare logic.
+/// If the payload is not dropped, the operation is pushed to the confirmation queue.
+/// If there are an error on retrieving payload ids or statuses for an operation, the operation
```

Comment:
In very simple terms, this function should decide if some messages should be sent straight to the confirm queue rather than filter. Given that, I would actually prepare the operation if retrieving the payload status fails, to keep things simple

Reply:
Done